sources/logdog/src/create_tarball.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+use crate::error;
+use crate::error::Result;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use flate2::write::GzEncoder;
+use flate2::Compression;
+use snafu::{ensure, ResultExt};
+use tar;
+
+/// Provides a function for compressing a directory's contents into a tarball.
+
+/// Creates a tarball with all the contents of directory `dir`.
+pub(crate) fn create_tarball<P1, P2>(dir: P1, outfile: P2) -> Result<()>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path>,
+{
+    // present an error if the output directory does not exist.
+    if let Some(outdir) = outfile.as_ref().parent() {
+        ensure!(
+            Path::new(&outdir).is_dir(),
+            error::NoOutputDirectory { path: &outdir }
+        )
+    }
+
+    // present an error if we would overwrite an existing file.
+    ensure!(
+        !Path::new(outfile.as_ref()).exists(),
+        error::OutputFileExists {
+            path: PathBuf::from(outfile.as_ref()),
+        }
+    );
+
+    // compress files and create the tarball.
+    let tarfile = File::create(outfile.as_ref()).context(error::TarballFileCreate {
+        path: outfile.as_ref(),
+    })?;
+    let encoder = GzEncoder::new(tarfile, Compression::default());
+    let mut tarball = tar::Builder::new(encoder);
+    tarball
+        .append_dir_all(crate::TARBALL_DIRNAME, dir.as_ref())
```

Comment:
```suggestion
        .append_dir_all(crate::TARBALL_DIRNAME, dir)
```
Most path-related functions should also take `AsRef<Path>`, like this, so you shouldn't need to call `as_ref` yourself unless you're using the path for some final purpose.

If you *are* calling `as_ref` yourself more than once or twice, it's simpler to `let x = x.as_ref();` once at the top of the function, like in this function with `outfile`.  The functions taking `AsRef<Path>` will still take it.  Would you mind doing that simplification?